### PR TITLE
-march=native is default. Hardware::CPU: Add Skylake [Linux]

### PR DESCRIPTION
**Pull request by sjackman**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

### CODE REVIEW

File: Library/Homebrew/extend/os/linux/hardware/cpu.rb [hunk 1]
```diff
@@ -36,12 +39,10 @@ def family
             :haswell
           when 0x3d, 0x47, 0x4f, 0x56
             :broadwell
-          when 0x5e
+          when 0x4e, 0x55, 0x5e, 0x8e, 0x9e
             :skylake
-          when 0x8e
-            :kabylake
           else
-            :dunno
```
Comment (MikeMcQuaid) [inline Library/Homebrew/extend/os/linux/hardware/cpu.rb]: `:dunno` has a special meaning elsewhere in the code; is `:unknown_` going to be handled correctly with those?

Comment (sjackman) [reply Library/Homebrew/extend/os/linux/hardware/cpu.rb]: https://github.com/Homebrew/brew/search?q=dunno
I see special treatment of `:dunno` here:
https://github.com/Homebrew/brew/blob/4897ec02d0f00c37a21ddaa0bebff6b83b27d027/Library/Homebrew/hardware.rb#L14
which seems unnecessary since `-march=native` is the default. I don't see any other special treatment of `:dunno`. Do you know of a case that you can point me to?

Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/linux/hardware/cpu.rb]: > which seems unnecessary since `-march=native` is the default

Only for `clang`, it seems. I think it'd be maybe worth handling the `:unknown*` case specifically in `optimization_flags`?

Comment (sjackman) [reply Library/Homebrew/extend/os/linux/hardware/cpu.rb]: I've changed the default optimization flags to `-march=native` in commit https://github.com/Homebrew/brew/pull/5012/commits/440a5e3e27aa39b7495d187be647caf4f98ffc73.

`optimization_flags.fetch` is called in three places:
https://github.com/Homebrew/brew/blob/f354160c7179ff0a615f2023393213736f8d6191/Library/Homebrew/extend/ENV/super.rb#L233-L237
https://github.com/Homebrew/brew/blob/f354160c7179ff0a615f2023393213736f8d6191/Library/Homebrew/extend/ENV/std.rb#L210
In the first two cases, `bottle_arch` and `oldest_cpu` are guaranteed to be in `optimization_flags`. In the last case, a default value of `DEFAULT_FLAGS` is provided.

File: Library/Homebrew/extend/ENV/super.rb [hunk 2]
```diff
@@ -235,7 +235,7 @@ def determine_optflags
       Hardware::CPU.optimization_flags.fetch(arch)
     elsif Hardware::CPU.intel? && !Hardware::CPU.sse4?
       Hardware::CPU.optimization_flags.fetch(Hardware.oldest_cpu)
-    elsif compiler == :clang
+    elsif ![:gcc_4_0, :gcc_4_2].include?(compiler)
```
[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: As far as I know, `-march=native` has always worked; at least, it has always worked for me. I don't know whether some versions of Apple GCC either didn't support it, or implemented it incorrectly. I'd suggest changing the default to `-march=native` for all compilers, and wait for a bug report with Apple GCC 4.0 to roll in, which I'd be surprised if I ever saw it.

This line in Stdenv makes me think that `-march=native` does not work with Apple GCC 4.0.
https://github.com/Homebrew/brew/blob/8f03ea7ca9f9452cd75ac9d6ec99f8f502cdf877/Library/Homebrew/extend/ENV/std.rb#L106

@mistydemeo Do Apple GCC 4.0 and 4.2 support `-march=native`?

File: Library/Homebrew/extend/ENV/std.rb [hunk 3]
```diff
@@ -108,7 +108,7 @@ def gcc_4_0
 
   def gcc_4_2
     super
-    set_cpu_cflags
+    set_cpu_cflags "-march=core2 -msse4"
```
Comment (sjackman) [inline Library/Homebrew/extend/ENV/std.rb]: @mistydemeo @MikeMcQuaid Is this change necessary? Does Apple GCC 4.2 support `-march=native`?

File: Library/Homebrew/extend/ENV/super.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: @mistydemeo @MikeMcQuaid Does Apple GCC 4.0 and 4.2 support `-march=native`?

[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/ENV/super.rb]: @sjackman 4.2 does not. I'd suggest confirming with the lowest supporting, packaged `gcc@` and being conservative to setting it to the lowest version you can confirm works.

[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: I've confirmed that GNU GCC 4.4 and up support `-march=native`, and GNU GCC 4.1 does not support `-march=native`. Apple GCC 4.2 does not support `-march=native`. I don't know of any distributions of Linux whose default compiler is GNU GCC 4.3. I suggest we ignore that case and move on. See https://github.com/Linuxbrew/brew/wiki/Bottles#glibc-version

File: Library/Homebrew/extend/ENV/std.rb (hunk 3 above)
Comment (sjackman) [reply Library/Homebrew/extend/ENV/std.rb]: Mike says that Apple GCC 4.2 does not support `-march=native`.

File: Library/Homebrew/extend/ENV/super.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/ENV/super.rb]: That's fine with me. For final paranoia: maybe just remove the `4.3` values from https://github.com/Homebrew/brew/blob/master/Library/Homebrew/compilers.rb#L3-L4 and then we're good to go.

[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: Done in commit 13816a56875225b65812817d548aec161d2395d9

[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: We use this idiom in a few places in Homebrew/core
```ruby
  ("4.3".."4.6").each do |n|
    fails_with :gcc => n
  end
```
Currently `fails_with :foobar` where `:foobar` is a compiler that's not known to Homebrew is ignored and gives no output/warning/error. Perhaps it should give a warning.

[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/ENV/super.rb]: > Perhaps it should give a warning.

👍 would be great. Mind doing that? Doesn't need to block this PR, though.

[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: If it gives a warning, would you rather…
1. Leave `gcc-4.3` as a known compiler to Homebrew/brew.
2. Remove references to `gcc-4.3` from Homebrew/core

[the remark this answers is not in the document]
Comment (sjackman) [reply Library/Homebrew/extend/ENV/super.rb]: Opened PR https://github.com/Homebrew/brew/pull/5034 to produce an error when the compiler argument is unknown.

[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/ENV/super.rb]: Thanks @sjackman, will discuss further there.
